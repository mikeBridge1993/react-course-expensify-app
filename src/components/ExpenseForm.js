import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component {
  constructor (props){
    super(props);
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return {
        description
      }
    });
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => {
      return {
        note
      }
    });
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      console.log(amount)
      this.setState(() => {
        return {
          amount
        }
      });
    }
    
  };
  
  onDateChange = (createdAt) => {
   
    if (createdAt){
      this.setState(() => {
        return {
          createdAt
        }
      });
    }
    
  };

  onFocusedChange = ({ focused }) => {
    this.setState(() => {
      return {
        calendarFocused: focused
      }
    });
  };

  onSubmit = (e) => {
    e.preventDefault();
    console.log(this.state.createdAt.valueOf())
    if (!this.state.description || !this.state.amount){
      this.setState(() => {
      return {
        error: " Please provide description and amount."
      }
      });
    } else {
      console.log("great success");
      this.setState(() => {
        return {
          error: ''
        }
      });

      this.props.onSubmit(
        {
          description: this.state.description,
          amount: parseFloat(this.state.amount, 10)*100,
          createdAt: this.state.createdAt.valueOf(),
          note: this.state.note
        }
      );
    }
  };

  render() {
    return (
      <div className="container-fluid expenses">
      {this.state.error && <div className="alert alert-danger"><strong>Error.</strong> {this.state.error}</div>}
      <form onSubmit={this.onSubmit} className="p-2 rounded border border-secondary">
        <div className="form-group w-90">
          <label>Description</label>
          <input type="text" className="form-control" 
          autoFocus 
          placeholder="Description" 
          value={this.state.description}
          onChange={this.onDescriptionChange}
          />
          <small className="form-text text-muted">Please describe your expense.</small>
        </div>
        <div className="form-group  w-90">
          <label>Amount</label>
          <input type="text" className="form-control" 
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
          />
        </div>
        <div className="form-group w-90">
          <label>Notes:</label>
          <textarea type="text" className="form-control" placeholder="Notes"
          value={this.state.note}
          onChange={this.onNoteChange}
          ></textarea> 
        </div>
        <div className="form-group w-90">
        <label>Date:</label><br />
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.calendarFocused} // PropTypes.bool
          onFocusChange={this.onFocusedChange} // PropTypes.func.isRequired
          numberOfMonths={1}
          isOutsideRange={() => { false }}
        />
        </div>
        <button type="submit" className="btn btn-primary">Submit</button>
        
      </form>
      </div>
    )
  };
}

