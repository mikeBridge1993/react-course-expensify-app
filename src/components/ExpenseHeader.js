import React from 'react';

const ExpenseHeader = () => {
  return (
    <div className="container-fluid expenses">
      <hr className="my-0 py-2"/>
      <h1 className="display-4 my-0 py-3">Expenses List</h1>
      <hr className="mt-2 py-2"/>
    </div>
  );
}



export default ExpenseHeader;
