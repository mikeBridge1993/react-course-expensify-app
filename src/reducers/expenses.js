const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(({id}) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((el) => {
          if(el.id === action.id){
            return {
              ...el,
              ...action.updates
            };
          }
          console.log("Not found");
          return el;
      });
    default:
      return state;
  }
}


