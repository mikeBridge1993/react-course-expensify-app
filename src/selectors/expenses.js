import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const startDateMatch =  startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy[0] === 'date'){
      if(sortBy[1] === 'high'){
        return a.createdAt < b.createdAt ? 1 : -1;        
      } else {
        return a.createdAt < b.createdAt ? -1 : 1;        
      }
    } else if (sortBy[0] === 'amount') {
       if(sortBy[1] === 'high'){
        return a.amount < b.amount ? 1 : -1;
      } else {
        return a.amount < b.amount ? -1 : 1;        
      }
    }
  });
};
