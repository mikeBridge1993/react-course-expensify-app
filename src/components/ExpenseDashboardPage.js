import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseHeader from './ExpenseHeader'

const ExpenseDashboardPage = () => (
  <div className="container-fluid expenses">
    <ExpenseHeader />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;