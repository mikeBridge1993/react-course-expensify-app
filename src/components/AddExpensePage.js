import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="container-fluid expenses">
          <hr className="my-0 py-2"/>
          <h1 className="display-4 my-0 py-3">Create New Expense</h1>
          <hr className="mt-2 py-2"/>
          <p className="lead">New expense:</p>  
        </div>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);