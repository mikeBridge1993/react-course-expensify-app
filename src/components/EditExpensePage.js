import React from 'react';
import { connect } from 'react-redux';
import {editExpense, removeExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="container-fluid expenses">
        <hr className="my-0 py-2"/>
        <h1 className="display-4 my-0 py-3">Edit Expense</h1>
        <hr className="mt-2 py-2"/>
        <p className="lead">Expense:</p>  
        </div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
        <div className="container-fluid expenses">
          <button onClick={this.onRemove}
            className="btn ml-0 btn-danger mt-2">
            Delete</button>
          
        </div>
      </div>
      
    );
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id;
      }
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);