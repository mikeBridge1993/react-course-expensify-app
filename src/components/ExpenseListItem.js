import React from 'react';
import { connect } from 'react-redux';
import {removeExpense} from '../actions/expenses';
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = (props) => {
  return (
    <li className="list-group-item d-flex justify-content-between">
      <div className="d-flex ExpenseListItem rounded">
          <span className="mr-auto text-secondary bold"><Link to={"/edit/" + props.id} >{props.description}</Link></span>
          <span className=" badge badge-pill badge-success mx-2">{props.amount}<i className="fa fa-euro ml-1"></i></span>
          <span className="badge badge-pill ml-auto badge-primary">{moment(props.createdAt).format("DD-MM-YYYY")}<i className="fa fa-calendar ml-1"></i></span>
      </div>
      <div>
      <button onClick={e => props.dispatch(removeExpense({id: props.id}))} className="delete-button  px-2 py-1"><i className="fa fa-times fa-1x delete-button--icon"></i></button>
      </div>
    </li>
  );
}

export default ExpenseListItem;



