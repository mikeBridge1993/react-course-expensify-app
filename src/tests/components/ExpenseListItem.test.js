import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseListItem} from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with expenses', () => {
  const expenseData = expenses[0];
  const wrapper = shallow(<ExpenseListItem {...expenseData} />);
  expect(wrapper).toMatchSnapshot();
});

