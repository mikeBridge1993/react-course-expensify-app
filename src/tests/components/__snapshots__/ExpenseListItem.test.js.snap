// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expenses 1`] = `
<li
  className="list-group-item d-flex justify-content-between"
>
  <div
    className="d-flex ExpenseListItem rounded"
  >
    <span
      className="mr-auto text-secondary bold"
    >
      <Link
        replace={false}
        to="/edit/1"
      >
        Gum
      </Link>
    </span>
    <span
      className=" badge badge-pill badge-success mx-2"
    >
      195
      <i
        className="fa fa-euro ml-1"
      />
    </span>
    <span
      className="badge badge-pill ml-auto badge-primary"
    >
      31-12-1969
      <i
        className="fa fa-calendar ml-1"
      />
    </span>
  </div>
  <div>
    <button
      className="delete-button  px-2 py-1"
      onClick={[Function]}
    >
      <i
        className="fa fa-times fa-1x delete-button--icon"
      />
    </button>
  </div>
</li>
`;
